### ISSUE TITLE
possible small issue with the link/lightbox plugin

### ISSUE THREAD

**Issue by ronm123** [thexerteproject/xerteonlinetoolkits#955]
@JohnSmith-LT while updating the release notes I think I found a small bug with the link/lightbox plugin.
I was adding images and setting those images to align center and trying to attach links to the images.
I think what I found (but wanted to test further) that this all works fine when the image is set to no alignment or left or right aligned but when set to centre the link url doesn't get saved and in fact gets removed if already added with the alignment set to one of the other options and then changed to center.

**Comment by JohnSmith-LT** [comment]
Hmm... that's a strange one. I'll look into it...

**Comment by JohnSmith-LT** [comment]
Hi @ronm123 

Have spent an hour trying to reproduce this and I did see it once but the rest of the time it is working normally on an image that is centre aligned... not really sure what you were seeing but can you detail the steps to reproduce as i've been tracing the plugin in and out with various alignments and it always picks up what is there and writes what i'm changing it to...

**Comment by ronm123** [comment]
Hi @JohnSmith-LT I wondered whether perhaps this was a difference between the show toolbar option vs just expanding the toolbar option like we had once before but it's not. Here's a recording and although this is my local xampp develop install I first noticed this on our xot.xerte install which is 3.9.
![image_link_test](https://user-images.githubusercontent.com/5167146/93938129-b5da8800-fd20-11ea-90f1-f317e3c16be4.gif)

**Comment by JohnSmith-LT** [comment]
Yes it's a strange one because I still cannot reproduce:

![TestLightboxCentreAlignedImage1](https://user-images.githubusercontent.com/3907362/94065239-c6ebcd80-fde2-11ea-8a1a-416054825876.gif)

**Comment by ronm123** [comment]
@JohnSmith-LT I've just been looking at this again and I can reproduce this consistently including on installations upgraded to 3.9 not just develop.
I tested on installations that are still 3.8.5-23 and they work fine but don't contain the lightbox plugin code.
I also tested by creating a new project and plain text page in 3.8.5-23 and then exported that and imported into 3.9 and with that plain text page in 3.9 I can't reproduce the issue. BUT if I then create a new plain text page in that same imported project I can reproduce the issue. Can you try that e.g. create a new plain text page in your existing project and also create a new project and new plain text page in that.
I'm a bit nervous about upgrading installations to 3.9 in case this breaks any links in existing projects where images are aligned center. I don't think existing projects are impacted until the page with the center aligned image and link are edited but we don't have enough people and projects being tested to make me confident about this. @torinfo @FayCross can you reproduce any of this?

**Comment by ronm123** [comment]
BTW happy to have a quick zoom to demo this.

**Comment by JohnSmith-LT** [comment]
Hi @ronm123 , I get your hesitation but if you need to upgrade then at the very least then you should be able to just disable the lightbox plugin and then it won't add in that functionality - just remove xotlightbox from https://github.com/thexerteproject/xerteonlinetoolkits/blob/3e348620c4c03f1c90c9fd4a847399df7b1fb8a1/editor/js/vendor/ckeditor/config.js#L26 .

It's really hard to debug this since I haven't been able to reproduce it but i'll try with bigger and smaller LOs - leave  it with me and i'll try again tonight. Regarding the importing a plain text page from an older install and that not having the issue - i can't see why that would be the case other than perhaps some timing issue with ckeditor but I will try that also.

**Comment by torinfo** [comment]
@ronm123 @JohnSmith-LT I cannot reproduce this on a local development install. Ron, you've got access to the same install (let me know if you don't have the creds anymore). Can you reproduce in that install?

BTW what tool do you use to create the screencaptures?

**Comment by ronm123** [comment]
@torinfo not sure what install you mean but I created a project named image link issue (id 230) on xot.xerte and you both have co-author access. I'm using Chrome and in that project you'll find a plain text page containing a line of text, then a new para, then an image aligned left with a link attached. 
In the editor select the image and then check the link - all should be fine.
Then double click the image and set alignment to centre.
Check the link again and it will be empty or at least that's consistently the case for me and I'm testing and typing this on a completely different laptop so I don't understand how you both can't reproduce it.

Re screen capture it's https://github.com/NickeManarin/ScreenToGif that @JohnSmith-LT recommended previously.

**Comment by JohnSmith-LT** [comment]
@torinfo I haven't been able to locate my login for xot.xerte for a long time. Could you resend please when you have some time...

@ronm123 Can you reproduce this on an up-to-date develop? I tried again and like the gif I sent, centre align still populates the link fine when I click the link button (both on menu and context menu) and also lets me change the url... wondering if it has something to do with the underlying HTML... Can you copy the HTML with the image aligned centre but BEFORE you click the link button? If when you click link it is indeed broken then can you please post the HTML either here or email it to me... 

It's the only thing that I can think of, that something in the HTML (either created before or after parsing by the plugin that is then causing the plugin to choke... You are not seeing any console errors either?

**Comment by ronm123** [comment]
@JohnSmith-LT yes I can reproduce on latest develop and 3.9. I can reproduce with Chrome and Firefox on different laptops too. I can reproduce consistently on xot.xerte.org. The html is simple as can be. Please try these exact steps:
Add a new plain text page
with the show toolbar ticked add a para after the default line of text and insert an image
with the image selected click the link icon and attach a link to the image
with the image selected click the link icon to verify that the link has saved
double click the image and select align center
with the center aligned image selected click the link icon to check the link - for me the link box is empty
unless I then change the image alignment to something other than center I can't re attach the link

But as I've said I have been able to attach links with images aligned center with an existing project/page.

**Comment by torinfo** [comment]
@ronm123 @JohnSmith-LT 
Iteresting. I can not reproduce it on your page Ron, but if I insert a new page above, than I can in the xot.xerte.org.uk install. I have not been able to reproduce on my develop install.

**Comment by JohnSmith-LT** [comment]
@ronm123 I have been able to replicate on develop! It is definitely related to the problem we had before and I can only reproduce if the Show Toolbar checkbox is ticked... previously I was noticing that when you tick that box the ids of all the ckeditor boxes on the page change to a different id yet the plugin seemed to hold onto the ref for the old editor... 

Can you confirm @ronm123 on one of your installs that it works for you if you DON'T tick that box? I know we need to have that working also but I may look first to see if there is a different way to have that functionality work without changing the ckeditor instance and id.... unless you have any ideas @torinfo ?

**Comment by ronm123** [comment]
@JohnSmith-LT @torinfo I have to confess I'm glad you've finally both been able to reproduce this - I wasn't making it up! ;-)
However to summarise what I've said in previous posts here:

- I can reliably replicate it on both develop and 3.9 and different browsers and different PC's.
- I can also replicate it with and without show toolbar checked.
- There are times when I then can't replicate it on an already added page (with and without show toolbox checked) but then if I add a new page to that same project I can reliably replicate it again on that new page.
- I first noticed this during a real world task e.g. populating the release notes where I wanted a centre aligned screenshot linking to an example and couldn't add the link. So obviously this happens in Bootstrap too.

So @JohnSmith-LT no I'm afraid your theory about this reflecting a difference between show toolbox checked or not doesn't seem to be the case for me. At times early on that did seem to be the case but then at one point it also seemed like it made a difference if the image was set to % width/height rather than left at initial dimensions but after further tests that didn't seem to matter either. I don't understand why you can't both replicate this as often or reliably as me but at least you have seen it happen. Like I said I'm happy to arrange a quick zoom to demonstrate

**Comment by JohnSmith-LT** [comment]
Update on this @ronm123 and @torinfo and it gets weirder and weirder because the link is stripped out BEFORE you even click the link button... Can you try to reproduce this one @ronm123 ?

So follow your instructions:

Add a new plain text page
with the show toolbar ticked add a para after the default line of text and insert an image
with the image selected click the link icon and attach a link to the image

Then, open dev and type this in the console... `CKEDITOR.instances["textarea_1"].getData()` and you should see something like:

```
<p>Enter text for the page here</p>

<p><a href="http://xerte.org.uk" target="_blank"><img alt="" height="90" src="http://localhost/PROJECTS/xot_(04-04-2020)/USER-FILES/31-guest2-Nottingham/media/Xerte-Logo-cropped.png" width="150" /></a></p>
```

Which is great! Then either double click the image and change to center alignment or click the centre alignment button on the toolbar.. Then run this again in dev console `CKEDITOR.instances["textarea_1"].getData()` and you should see something like:

```
<p>Enter text for the page here</p>

<p></p>

<p style="text-align:center"><img alt="" height="90" src="http://localhost/PROJECTS/xot_(04-04-2020)/USER-FILES/31-guest2-Nottingham/media/Xerte-Logo-cropped.png" width="150" /></p>

<p></p>
```

So basically it's not the Link plugin and i'm off the hook ;-)

**Comment by JohnSmith-LT** [comment]
So @ronm123 on a hunch I re-wrote the Show Toolbar code to just toggle the editors that were in the wrong state and now miraculously I cannot reproduce this using your steps @ronm123 or any others that I can think of...

Can you please test this at your end and @torinfo can you cast an eye over the Show Toolbar code that i've changed to see if you think that will interfere with anything else in the editor?

Hope this is fixed as it's driving me crazy!

**Comment by ronm123** [comment]
@JohnSmith-LT @torinfo A few comments...
1. I know this is a nightmare and very weird especially as there seems to be so many variations of being able to reproduce and then not being able to reproduce.
2. I've just be testing for about an hour and notice additional weird behaviour but also something that I can consistently reproduce too.
3. @JohnSmith-LT yes I can reproduce the dev console steps where just aligning centre removes the link and apologies I never meant to suggest that this was definitely due to the lightbox plugin and the title should have been "issue possibly due..." rather than "possible issue due..."
4. It does get weirder too and I can still reproduce it having pulled down the changes and cleared cache but only when crating a new page each time. e.g. I've ended up with a project with loads of plain text pages and consistent results:
Create a new plain text page
Hit return, add an image, leave it no alignment and add a link then check the link has saved - all good.
Double click the image and select Align Centre - the link disappears and also I notice an extra <p></p> get added above the image.
Change the alignment to left or  none, re add the link, publish and close the editor.
Re open in the editor, turn toolbar on, try to reproduce by align centre but can't link remains and can't reproduce.
Add a new plain text page and repeat everything the same e.g. link disappears on align center but only on that new page.

I also noticed another weird issue but perhaps this has always been the case and I just haven't noticed...
I'm now testing with a small image that is 64px x 73px (attached). If I leave that at fixed dimensions then regardless of what alignment I set it displays the same size in the editor and the player. If I change it to say 30% in the editor then when alignment is none, left or right it's the same size in the editor but if align is set to center it becomes much much smaller. When playing the project the % size set is the same regardless of alignment but it's still a bit odd why there is this difference in the editor when center align is set. Also align right displays different in the editor if the images is % or fixed size e.g. when %  it isn't against the right edge but when fixed size it is.
![journeys](https://user-images.githubusercontent.com/5167146/95013326-7e7cad00-0637-11eb-924f-9bfef43ef126.png)

**Comment by ronm123** [comment]
@JohnSmith-LT @torinfo just a note to say I edited that previous summary to remove some typos and make it clearer with some additional punctuation etc.

**Comment by JohnSmith-LT** [comment]
Hi @ronm123 , I was just jesting about being off the hook... It is interesting though that during my, probably inferior, testing I could only reproduce the link being stripping out by the centre align button ONLY IF the lightbox plugin was enabled and not without it...

Hmm, i'll definitely give it another go with your suggested steps... as you say I think some of these things have perhaps been there for a while but they only seem to resurface during very specific circumstances...

**Comment by ronm123** [comment]
@JohnSmith-LT sorry if I missed that instruction but to be clear I wasn't testing with the lightbox plugin disabled only with it enabled. I did test on an earlier Xerte version before the plugin existed and couldn't reproduce the error on that but that was early on in all this testing so can revisit that if needed or likewise test locally with the plugin disabled if needed too.

**Comment by JohnSmith-LT** [comment]
No leave it with me again @ronm123 and i'll see what I can come up with - I have a few more ideas to try to see if I can figure out what is going on... I googled too and don't see anything online about the centre justify image stripping out links so it must be local to us but I just find it very strange that it does it without even clicking on the plugin again - although i'm working on the assumption that when you add the link initially then possibly something i'm doing in code is modifying the DOM sufficiently that while "Source" mode shows the correct HTML, the underlying DOM nodes are somehow different...
